Rendering optimization and refactor

RiugaBachi opened the pull request:
:reimumug:

Sorry for the delays, kept my word in the end since I am (finally) feeling a bit better today.

e8439932fcd87cbc9ada26a410920aac4275b794 contains a perfectly good IntMap-only implementation. I'm happy to submit another pull request tomorrow for a benchmark suite if that is something you would like to see. If there's ever a need for comparison between that and the current optimized approach, I would just branch from that commit and merge benchmarks + test suite.

RiugaBachi committed:
Swapped out Array for lazy IntMap

RiugaBachi committed:
Implemented hybrid IntMap + association list source block traversals.

RiugaBachi committed:
Refactored Errata.Internal.Render, added optimization documentation.

RiugaBachi committed:
Merge branch 'master' of https://github.com/1computer1/errata into master
Merged test suite.

RiugaBachi committed:
Fixed edge case failures in test suite.

- Made `Monoid` a superclass of `Source`.
- Every slice from `slices` now infinitely tails `mempty` to support
  out-of-bounds pointers.

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -1,6 +1,8 @@
+{-# LANGUAGE OverloadedStrings   #-}
+{-# LANGUAGE RecordWildCards     #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE ViewPatterns #-}
```
Align the `#-}`.

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -36,50 +38,110 @@ renderErrors :: Source source => source -> [Errata] -> TB.Builder
+        blocks = errataBlock : errataBlocks
+
+        -- The pointers grouped by line.
+        blockPointersGrouped = I.fromListWith (<>) . map (pointerLine &&& pure) . blockPointers <$> blocks
```
I don't think the `Control.Arrow` import is justified for just this one use of `(&&&)` :P

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -21,10 +23,10 @@ module Errata.Internal.Render
 import           Data.List
-import qualified Data.List.NonEmpty as N
-import qualified Data.Map.Strict as M
+import qualified Data.IntMap as I
```
Put this in alphabetical order please! One day I'll have a formatter I like and enforce it, but not today.

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -36,50 +38,110 @@ renderErrors :: Source source => source -> [Errata] -> TB.Builder
+renderErrata :: Source source => [source] -> Errata -> TB.Builder
 renderErrata slines (Errata {..}) = errorMessage
     where
+        -- | Header block + body blocks
```
`--` instead of `-- |`

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -36,50 +38,110 @@ renderErrors :: Source source => source -> [Errata] -> TB.Builder
+        -- Turns a list into a list of tail slices of the original list,
+        -- each element at index @i@ dropping the first @i@ elements of the original list.
+        slices [] = repeat (repeat mempty)
+        slices xs = (xs <> repeat mempty) : slices (tail xs)
```
Are the `mempty`s here just `[]`? Why the need to repeat infinitely? I'm guessing to allow out-of-bounds access? Do add a comment and a type signature.

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -98,50 +160,47 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
         -- The next line is a line we have to decorate with pointers.
-        makeSourceLines _ (n:ns)
-            | Just p <- M.lookup n pointersGrouped = makeDecoratedLines p <> makeSourceLines 0 ns
+        makeSourceLines _ (pr@(n,_):ns)
```
Should add a space after the `,`

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -98,50 +160,47 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
 
         -- The next line is an extra line, within a limit (currently 2, may be configurable later).
-        makeSourceLines extra (n:ns)
+        makeSourceLines extra ((n,l):ns)
```
^

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -151,69 +210,69 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
                 -- There is a single extra line, so we can use that as the before-line.
                 -- No need for omission, because it came right before.
-                ([n], _) ->
+                ([(n,l)], _) ->
```
^

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -151,69 +210,69 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
                 -- We use the last one as the before-line.
                 (_, _) ->
-                    let n = last es
+                    let (n,l) = last es
```
^

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -151,69 +210,69 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
-        makeDecoratedLines :: N.NonEmpty Pointer -> [TB.Builder]
-        makeDecoratedLines pointers = (linePrefix line <> TB.fromText lineConnector <> sline) : decorationLines
+        makeDecoratedLines :: [Pointer] -> (Line, source) -> [TB.Builder]
+        makeDecoratedLines pointers (num,line) = 
```
^

1Computer1 commented on src/Errata/Internal/Render.hs:
```diff
@@ -151,69 +210,69 @@ renderSourceLines slines (Block {..}) lspans = unsplit "\n" sourceLines
                 decorationLines = if
                     -- There's only one pointer, so no need for more than just an underline and label.
-                    | N.length pointersSorted' == 1                           -> [underline pointersSorted']
+                    | length pointersSorted' == 1                             -> [underline pointersSorted']
```
This is pretty big spacing, so I think you should just remove the horizontal alignment here.

1Computer1 commented on src/Errata/Source.hs:
```diff
@@ -21,7 +21,7 @@ A class for manipulating and converting source text.
 For @ByteString@ source types, you should convert it to one of the built-in instances with your encoding of choice.
 -}
-class Source s where
+class Monoid s => Source s where
```
You should probably document what is expected of the `Monoid` requirement.

1Computer1 commented:
A benchmark suite would be very nice if you have the time. Would it be the adhoc one you used, or an actual suite with `criterion`?

RiugaBachi committed:
Applied requested changes.

- Imports are now in alphabetical order
- Tuple elements are now spaced
- Pragmas are now aligned
- Documented `slices` semantics
- Added a type signature for `slices`
- Documented `Monoid` superclass requirement for `Source`
- Removed redundant whitespace on long branching expressions

RiugaBachi commented:
A comprehensive one with `criterion`.

1Computer1 commented:
That sounds great then. I'll take one last look over this in the morning and merge it then.
